rls/src/test/java/io/grpc/rls/AdaptiveThrottlerTest.java

```diff
@@ -29,82 +30,96 @@
+    assertThat(throttler.throttledStat.get(fakeTicker.read()))
         .isEqualTo(1L);
-    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+    assertThat(throttler.getThrottleProbability(fakeTicker.read()))
         .isWithin(TOLERANCE)
         .of(1.0f / 3.0f);
 
     // Skip to half second mark from the beginning (half the duration).
-    fakeClock.forwardTime(500 - (fakeClock.currentTimeMillis() - startTime), TimeUnit.MILLISECONDS);
+    fakeClock.forwardTime(500 - (fakeClock.currentTimeMillis() - startTime),
+        TimeUnit.MILLISECONDS);
 
     // Request 3, throttled by backend
     assertThat(throttler.shouldThrottle(0.4f)).isFalse();
     fakeClock.forwardTime(1L, TimeUnit.MILLISECONDS);
     throttler.registerBackendResponse(true);
 
-    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(3L);
-    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(2L);
-    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+    assertThat(throttler.requestStat.get(fakeTicker.read())).isEqualTo(3L);
+    assertThat(throttler.throttledStat.get(fakeTicker.read())).isEqualTo(2L);
+    assertThat(throttler.getThrottleProbability(fakeTicker.read()))
         .isWithin(TOLERANCE)
         .of(2.0f / 4.0f);
 
     // Request 4, throttled by client.
     assertThat(throttler.shouldThrottle(0.4f)).isTrue();
     fakeClock.forwardTime(1L, TimeUnit.MILLISECONDS);
 
-    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(4L);
-    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(3L);
-    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+    assertThat(throttler.requestStat.get(fakeTicker.read())).isEqualTo(4L);
+    assertThat(throttler.throttledStat.get(fakeTicker.read())).isEqualTo(3L);
+    assertThat(throttler.getThrottleProbability(fakeTicker.read()))
         .isWithin(TOLERANCE)
         .of(3.0f / 5.0f);
 
     // Skip to the point where only requests 3 and 4 are visible.
     fakeClock.forwardTime(
         1250 - (fakeClock.currentTimeMillis() - startTime), TimeUnit.MILLISECONDS);
 
-    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(2L);
-    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(2L);
-    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+    assertThat(throttler.requestStat.get(fakeTicker.read())).isEqualTo(2L);
+    assertThat(throttler.throttledStat.get(fakeTicker.read())).isEqualTo(2L);
+    assertThat(throttler.getThrottleProbability(fakeTicker.read()))
         .isWithin(TOLERANCE)
         .of(2.0f / 3.0f);
   }
+
+  @Test
+  public void negativeTickerValues() {
+    fakeClock.forwardTime(-300, TimeUnit.MILLISECONDS);
```

Comment:
Seems that this should be based on the current value of fakeClock and not assume what it is initialized to.

Reply:
This is specifically checking that when fakeClock has negative values in this range that the slot identification logic still works correctly.

Reply:
Added a description for new test so that its purpose is clearer.

Reply:
I understood the purpose. But how do you know "fakeClock has negative values" for this test? I suggest getting the current nanoTime from fake clock and then rewinding time based on it, instead of hard-coding 300ms.

Some day I want fake clock to choose a random nanoTime or Integer.MAX_VALUE in its constructor.